handler.go

```diff
@@ -0,0 +1,343 @@
+	}
+
+	if err == io.EOF {
+		err = closeSendStream(ctx, stream)
+	}
+	if err != nil {
+		return err
+	}
+
+	res, err := receiveStreamMessage(ctx, stream)
+	if err != nil {
+		return err
+	}
+	return responseHandlerFn(res)
+}
+
+// makeBidiStream starts bi-directional streaming rpc
+func makeBidiStream(ctx context.Context, stream *yarpctransport.ClientStream,
+	nextBodyFn streamRequestSupplierFn, responseHandlerFn streamResponseHandlerFn) error {
+
+	var wg sync.WaitGroup
+	var sendErr error
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	wg.Add(1)
+	// Start go routine to concurrently send stream messages.
+	go func() {
+		defer wg.Done()
+
+		var err error
+		for err == nil {
+			var reqBody []byte
+			reqBody, err = nextBodyFn()
+			if err == io.EOF {
+				err = closeSendStream(ctx, stream)
+				break
+			}
+			if err != nil {
+				// Cancel the context to unblock the routine waiting on receiving
+				// stream messages.
+				cancel()
+				break
+			}
+
+			err = sendStreamMessage(ctx, stream, reqBody)
+		}
+
+		if err != nil {
+			sendErr = err
+		}
+	}()
+
+	var receiveErr error
+	for receiveErr == nil {
+		var resBody []byte
+		resBody, receiveErr = receiveStreamMessage(ctx, stream)
+		if receiveErr != nil {
+			break
+		}
+
+		receiveErr = responseHandlerFn(resBody)
+	}
+
+	cancel()
+	wg.Wait()
+
+	if sendErr != nil && sendErr != io.EOF {
+		return sendErr
+	}
+
+	if receiveErr != nil && receiveErr != io.EOF {
+		return receiveErr
+	}
+
+	return nil
+}
+
+// sendStreamMessage sends the stream message using message body provided
+func sendStreamMessage(ctx context.Context, stream *yarpctransport.ClientStream, msgBody []byte) error {
+	// TODO: print the stream message being sent on STDOUT to inform user about
+	// request message dispatch.
+	req := &yarpctransport.StreamMessage{Body: ioutil.NopCloser(bytes.NewReader(msgBody))}
+	if err := stream.SendMessage(ctx, req); err != nil {
+		return fmt.Errorf("Failed while sending stream request: %v", err)
+	}
+	return nil
+}
+
+// receiveStreamMessage receives and returns the message from the given stream
+func receiveStreamMessage(ctx context.Context, stream *yarpctransport.ClientStream) ([]byte, error) {
+	msg, err := stream.ReceiveMessage(ctx)
+	if err == io.EOF {
+		return nil, err
+	}
+	if err != nil {
+		return nil, fmt.Errorf("Failed while receiving stream response: %v", err)
+	}
+
+	bytes, err := ioutil.ReadAll(msg.Body)
+	if err != nil {
+		return nil, fmt.Errorf("Failed while reading stream response: %v", err)
+	}
+	return bytes, err
+}
+
+// closeSendStream closes the stream from the client side while
+// stream can continue to receive messages from server
+func closeSendStream(ctx context.Context, stream *yarpctransport.ClientStream) error {
+	// YARPC stream.Close method internally invokes closeSend on gRPC clientStream.
+	if err := stream.Close(ctx); err != nil {
+		return fmt.Errorf("Failed to close send stream: %v", err)
+	}
+	return nil
+}
```

Comment:
Ideally we could have a files `handler.go`, `handler_unary.go` and `handler_stream.go` here, this will simplify code maintenance

Reply:
Since the methods of streaming and unary belong to `requestHandler`, I prefer keeping them in the same file. And I think handler.go isn't very huge either